src/librustc_typeck/check/demand.rs

```diff
@@ -58,7 +59,68 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let suggestions = self.check_ref(expr, checked_ty, expected);
+            let mut err = self.report_mismatched_types(origin, expected, expr_ty, e);
+            if let Some(suggestions) = suggestions {
+                err.help(&suggestions);
```

Comment:
There is a `span_suggestion` method on `DiagnosticBuilder`:

```rust
    /// Prints out a message with a suggested edit of the code.
    ///
    /// See `diagnostic::RenderSpan::Suggestion` for more information.
    pub fn span_suggestion<S: Into<MultiSpan>>(&mut self,
                                               sp: S,
                                               msg: &str,
                                               suggestion: String)
                                               -> &mut Self {
```

I think the idea is that you replace the given span with the text `suggestion`, and also display the help message `msg`. Something like:

```
err.span_suggestion(span, "consider using the `&` operator", format!("&{}", span_to_string(span))
```